### PR TITLE
fix(chart-data-api): improve chart data endpoint errors

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY

TL;DR: this makes viz plugins behave in the same way as legacy viz plugins when the request fails or returns no data.

This fixes the following problems in the chart data endpoint:
- if no data was returned, the chart would not render, leaving the old chart on screen, potentially even an incorrect chart type. This was due to the legacy endpoint returning a 200 with empty data, as opposed to the new endpoint returning a 200 response with an error field but without a data field. To harmonize behavior, `QueryContext` was updated to return a non-error response with an empty data field for empty result sets.
- The `error_message` property of `QueryResult` was not being populated for failed requests in SQLA model, leading to the `error` field on `QueryObject` always being `None`.
- Errors were not picked up from `QueryObject`s when calling the chart data endpoint. This meant that erroneous requests were returned as 200s with an error-field that was not handled in any way, causing the chart to render in unexpected ways. Now the `error` field is is checked for each response dataset; if a single one is defined, a 400 is returned (these are mostly expected to be incorrect SQL).
- The error message from erroneous requests (=incorrect request schema) were not displayed, making it difficult to see in the frontend what the problem was.

### AFTER
When no data is returned, the chart data endpoint returns a response that the `Chart` component can properly identify as an empty response. Previously this left the previous rendered state visible.
image

When an incorrect request is performed, the error message is displayed, making debugging easier. Previously only "An error occurred." was displayed (please note that the error in this screenshot has been fixed by #10299).
image

When an a query is incorrect, the error is displayed in the warning (previously this would print "An error occurred while rendering the visualization: TypeError: Cannot read property 'map' of undefined") 
image

### TEST PLAN
Local testing.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/charts/api.py [hunk 1]
```diff
@@ -465,13 +465,16 @@ def data(self) -> Response:
...
+                message=_("Request is incorrect: %(error)s", error=error.messages)
             )
         try:
             query_context.raise_for_access()
         except SupersetSecurityException:
             return self.response_401()
         payload = query_context.get_payload()
+        for query in payload:
+            if query["error"]:
+                return self.response_400(message=f"Error: {query['error']}")
```
Comment (dpgaspar) [inline superset/charts/api.py:477]: Would be great if `query_context.get_payload()` would raise custom superset exceptions instead of checking for errors on the payload. But we can address that later

**Review by dpgaspar — APPROVED** [review]

File: superset/charts/api.py (hunk 1 above)
Comment (villebro) [reply superset/charts/api.py:477]: That's a great idea 👍  To unblock `0.37` I'll merge this as-is, but will improve exception handling after we ship the release.
